File: Data Structures & Algorithms/sqrtx/submission-3.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x == 0:
            return 0
        
        l, r = 1, (x//2)+1
        cur: int
        closest: int
        while l <= r:
            cur = l + (r-l) // 2
            res = cur * cur
            if res < x:
                l = cur + 1
                closest = cur
            elif res > x:
                r = cur - 1
            else:
                closest = cur
                break
        
        return closest

File: Data Structures & Algorithms/combination-target-sum/submission-3.cpp
class Solution {
public:

    void algo(vector<int>& nums, vector<int> current, vector<vector<int>>& res, int i, int curNum, int target){
        if(curNum == target){
            res.emplace_back(current);
            return;
        }

        if(curNum > target){
            return;
        }

        for(int x = i; x < nums.size(); x++){
            curNum += nums[x];
            current.emplace_back(nums[x]);
            algo(nums, current, res, x, curNum, target);
            curNum -= nums[x];
            current.pop_back();
        }

    }

    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        algo(nums, {}, res, 0, 0, target);

        return res;
    }
};
